src/hotspot/share/opto/subnode.cpp

```diff
@@ -1928,6 +1928,38 @@ bool BoolNode::is_counted_loop_exit_test() {
   return false;
 }
 
+template<typename IntegerType>
+static const IntegerType* integral_abs_value(const IntegerType* t) {
+  typedef typename IntegerType::NativeType NativeType;
+
+  assert(!t->is_con(), "Constant types must have already been handled");
+
+  NativeType lo_abs = uabs(t->_lo);
+  NativeType hi_abs = uabs(t->_hi);
```

Comment:
Converting unsigned to signed is C++ Undefined Behavior, is it not?

Reply:
This is a great point, I believe you're correct that it's UB. We currently do the same logic in the old code as well:
https://github.com/openjdk/jdk/blob/a0677d94d8c83a75cee054700e098faa97edca3c/src/hotspot/share/opto/subnode.cpp#L1945-L1947
However, I'm unsure what the best way to solve this would be. Do you happen to have any ideas? Thanks a lot!

Reply:
@dean-long  ?

Reply:
No converting unsigned to signed is not UB, the behaviour is the same as in Java.

Reply:
I believe it's actually implementation-defined, not UB, until C++ 20, according to discussion in this other PR:
https://github.com/openjdk/jdk/pull/24184#discussion_r2011464234